Spatial_Engine/Query/OpeningCurve.cs

```diff
@@ -0,0 +1,184 @@
+            if (opening.SpacerHeight == 0)
+            {
+                ctrlPts = new List<Point>
+                {
+                    new Point{ X = opening.WidthWebPost/2, Y = -opening.Height/2 },
+                    new Point{ X = opening.Width/2},
+                    new Point{ X = opening.WidthWebPost/2, Y = opening.Height/2},
+                    new Point{ X = -opening.WidthWebPost/2, Y = opening.Height/2 },
+                    new Point{ X = -opening.Width/2},
+                    new Point{ X = -opening.WidthWebPost/2, Y = -opening.Height/2 },
+                    new Point{ X = opening.WidthWebPost/2, Y = -opening.Height/2 },
+                };
+            }
+            else
+            {
+                double h = opening.Height + opening.SpacerHeight;
+                ctrlPts = new List<Point>
+                {
+                    new Point{ X = opening.WidthWebPost / 2, Y = - h / 2 },
+                    new Point{ X = opening.Width/2, Y = - opening.SpacerHeight/2},
+                    new Point{ X = opening.Width/2, Y = opening.SpacerHeight/2},
+                    new Point{ X = opening.WidthWebPost/2, Y = h/2},
+                    new Point{ X = -opening.WidthWebPost/2, Y =h/2 },
+                    new Point{ X = -opening.Width/2, Y = opening.SpacerHeight/2},
+                    new Point{ X = -opening.Width/2, Y = -opening.SpacerHeight/2},
+                    new Point{ X = -opening.WidthWebPost/2, Y = -h/2 },
+                    new Point{ X = opening.WidthWebPost/2, Y = -h/2 },
+                };
+            }
+
+            return new Polyline { ControlPoints = ctrlPts };
+        }
+
+        /***************************************************/
+
+        [Description("Returns the curve of a single opening in the XY plane.")]
+        [Input("opening", "The cellular opening to fetch the opening curve from.")]
+        [Output("curve", "The outline curve of a single opening.")]
+        public static PolyCurve OpeningCurve(this SinusoidalOpening opening)
+        {
+            if (opening == null)
+            {
+                Base.Compute.RecordError("Unable to query the curve from a null opening.");
+                return null;
+            }
+
+            Point basePt = new Point() { X = opening.WidthWebPost / 2, Y = -opening.Height / 2 };
+
+            NurbsCurve s1 = ApproxSine(basePt, opening.Height / 2, opening.SinusoidalLength);
+            NurbsCurve s2 = s1.Mirror(Plane.XZ).Flip();
+            NurbsCurve s3 = s2.Mirror(Plane.YZ).Flip();
+            NurbsCurve s4 = s3.Mirror(Plane.XZ).Flip();
+
+            return new PolyCurve
+            {
+                Curves = new List<ICurve>
+                {
+                    s1,
+                    s2,
+                    new Line{ Start = s2.EndPoint(), End = s3.StartPoint() },
+                    s3,
+                    s4,
+                    new Line{ Start = s4.EndPoint(), End = s1.StartPoint() }
+                }
```

Comment:
Happy that the height is correct - but the length/width is not: (this is for a width of 0.6m).
![image](https://user-images.githubusercontent.com/30078587/224297003-b79bc271-70d7-483a-af73-84eed2f96fea.png)

Reply:
Might need to tweak descriptions/inputs here. The length refers to the length of the sinusoidal part, not the full length.

Happy to chat through how we can make this clearer.

Reply:
Yeah - I guess you really want to be able to set the total width.

Reply:
Updated the create method do do this now